exitAfterDefer: false positive

exitAfterDefer seems to give a false positive with this code:

```
package main

import (
	"fmt"
	"io"
	"os"
)

func main() {
	items, err := process()
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n\n", err)
		defer os.Exit(1)
	} else {
		defer os.Exit(0)
	}

	fmt.Println("output", items)
}

func process() ([]int, error) { return []int{1, 2, 3}, io.EOF }
```

Output:
```
example> gocritic check .
./main.go:15:9: exitAfterDefer: os.Exit clutters `defer os.Exit(1)`
```

--- a/checkers/exitAfterDefer_checker.go
+++ b/checkers/exitAfterDefer_checker.go
@@ -52,7 +52,18 @@ func (c *exitAfterDeferChecker) VisitFuncDecl(fn *ast.FuncDecl) {
 		case *ast.DeferStmt:
 			deferStmt = n
 		case *ast.CallExpr:
+			// See #995. We allow `defer os.Exit()` calls
+			// as it's harder to determine whether they're going
+			// to clutter anything without actually trying to
+			// simulate the defer stack + understanding the control flow.
+			// TODO: can we use CFG here?
+			if _, ok := cur.Parent().(*ast.DeferStmt); ok {
+				return true
+			}
 			if deferStmt != nil {
+				if deferStmt.Call == n {
+					return true
+				}
 				switch qualifiedName(n.Fun) {
 				case "log.Fatal", "log.Fatalf", "log.Fatalln", "os.Exit":
 					c.warn(n, deferStmt)

--- a/checkers/testdata/exitAfterDefer/negative_tests.go
+++ b/checkers/testdata/exitAfterDefer/negative_tests.go
@@ -4,6 +4,18 @@ import (
 	"os"
 )
 
+func deferExitBranches(cond bool) {
+	if cond {
+		defer os.Exit(1)
+	} else {
+		defer os.Exit(0)
+	}
+}
+
+func deferExit() {
+	defer os.Exit(1)
+}
+
 func exitInsideLambda() {
 	defer println("before return")
 	// Exit inside some anonymous function is fine.

